travis/publishFolderToArtifactory.js

```diff
@@ -0,0 +1,138 @@
+/**
+ * This script deploys the contents of the source directory to the WorldWind content server at Ames. The
+ * content server uses Artifactory for managing the assets.
+ *
+ * Provide the Artifactory API key, directory of assets to upload, and root deployment path on the Artifactory instance.
+ */
+
+var crypto = require("crypto");
+var fs = require("fs");
+var http = require("https");
+var os = require("os");
+var path = require("path");
+var recursive = require("recursive-readdir");
+
+// the Artifactory instance API key
+var apiKey;
+
+/**
+ * Submits all of the files (recursively) in the provided source directory to the Artifactory server.
+ * @param sourceDirectory the directory containing files to deploy to the Artifactory instance
+ * @param destinationPath the root target path for the assets
+ */
+var deployDirectory = function (sourceDirectory, destinationPath) {
+    recursive(sourceDirectory, function (err, files) {
+        if (err) {
+            console.error(err);
+        }
+
+        for (var i = 0, len = files.length; i < len; i++) {
+            // remove the root directory path from the file path and convert to unix style slashes
+            var normalizedFilepath = path.normalize(files[i]).replace(path.normalize(sourceDirectory), "").replace(/\\/g, "/");
+            // format the path to conform to Artifactory path requirements
+            var normalizedPath = path.normalize("/" + destinationPath + "/" + normalizedFilepath);
+            deployFile(files[i], normalizedPath);
+        }
+    });
+};
+
+/**
+ * Deploys a file to the Artifactory server. Calculates the checksums and then deploys the file to the provided
+ * destination path. If there is an unsuccessfuly deployment (as defined by a non-201 HTTP status code) the script will
+ * exit.
+ * @param file the file path of the asset to deploy
+ * @param destinationPath the fully defined target deployment path on the Artifactory instance
+ */
+var deployFile = function (file, destinationPath) {
+
+    var options = generateOptions(file, destinationPath);
+    var req = http.request(options, function (res) {
+        var chunks = [];
+
+        res.on("data", function (chunk) {
+            chunks.push(chunk);
+        });
+
+        res.on("end", function () {
+            if (res.statusCode !== 201) {
```

Comment:
Don't we want res.statusCode >= 200 && res.statusCode < 400 ?

Reply:
Is it possible that the HTTP status is something other than 201 if the file is being replaced?

Reply:
Permissions server side prevent replacement/override, this could be expanded to your suggested
`res.statusCode >= 200 && res.statusCode < 400` if you think it adds generality to the script.

Reply:
It looks like Artifactory returns 201 for replacement/override as well. Would it be better to expand the status code anyway?

Reply:
@pdavidc I think it would be best to stick with the 201 status code being the only valid response.

The object of the function, and the API function it is targeting, is to create/deploy an asset. The behavior of the API to this point has been to return a 201 on a successful deployment. I hesitate to accept a broader range of status codes as I'm not sure we understand the full implications of a different status code.